capirca/lib/nsxt.py

```diff
@@ -0,0 +1,320 @@
+  'name',
+  'action',
+  'resource_type',
+  'display_name',
+  'marked_for_delete',
+  'overridden',
+  'rule_id',
+  'sequence_number',
+  'sources_excluded',
+  'destinations_excluded',
+  'source_groups',
+  'destination_groups',
+  'services',
+  'profiles',
+  'logged',
+  'scope',
+  'disabled',
+  'notes',
+  'direction',
+  'tag',
+  'ip_protocol',
+  'is_default',
+  'protocol',
+  'destination_port',
+  'source_address',
+  'destination_address',
+  'source_port'
+]
+
+
+# generic error class
+class Error(Exception):
+  """Generic error class."""
+  pass
+
+
+class UnsupportedNsxtAccessListError(Error):
+  """Raised when we're give a non named access list."""
+  pass
+
+
+class NsxtAclTermError(Error):
+  """Raised when there is a problem in a nsxt access list."""
+  pass
+
+
+class NsxtDuplicateTermError(Error):
+  """Raised when there is a duplicate."""
+  pass
+
+
+class NsxtUnsupportedCriteriaOperator(Error):
+  """Raised when an unsupported criteria comparison operator is encountered."""
+  pass
+
+
+class Term(aclgenerator.Term):
+  """Creates a  single ACL Term for NSX-T."""
+
+  def __init__(self, term, filter_type, applied_to=None, af=4):
+    self.term = term
+    # Our caller should have already verified the address family.
+    assert af in (4, 6)
+    self.af = af
+    self.filter_type = filter_type
+    self.applied_to = applied_to
+
+  def __str__(self):
+    """Convert term to a rule string.
+
+    Returns:
+      A rule as a string.
+
+    """
+    # Verify platform specific terms. Skip whole term if platform does not
+    # match.
+    if self.term.platform:
+      if 'nsxt' not in self.term.platform:
+        return ''
+    if self.term.platform_exclude:
+      if 'nsxt' in self.term.platform_exclude:
+        return ''
+
+    term_keywords = self.term.__dict__
+    unsupported_keywords = []
+    for key in term_keywords:
+      if term_keywords[key]:
+        if ('translated' not in key) and (key not in _NSXT_SUPPORTED_KEYWORDS):
+            unsupported_keywords.append(key)
+    if unsupported_keywords:
+      logging.warning('WARNING: The keywords %s in Term %s are not supported '
+                      'in Nsxt ', unsupported_keywords, self.term.name)
+
+    name = self.term.name
+
+    notes = ''
+    if self.term.comment:
+      notes = self.term.comment
+
+    action = 'ALLOW'
+    if self.term.action:
+      action = self.term.action
+
+    source_address = 'ANY'
+    if self.term.source_address:
+      source_address = []
+      for i in self.term.source_address:
+        source_address.append(str(i))
+
+    destination_address = 'ANY'
+    if self.term.destination_address:
+      destination_address = []
+      for i in self.term.destination_address:
+        destination_address.append(str(i))
+
+# Intentionally left with defaul values as these fields are mandatory in the API. If they need to be configured these should be extended to br read from the policy files.
+    services = 'ANY'
```

Comment:
In NSX-V, the XML for a service is output like this:

```python
          services.append(self._ServiceToString(proto,
                                                source_port,
                                                destination_port,
                                                icmp_types))
```

The NSX-T code does not seem to be using `source_port`, `destination_port`, `icmp_types`?